flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionService.java

```diff
@@ -32,87 +29,49 @@
  */
 public class TestingLeaderElectionService implements LeaderElectionService {
 
-    private LeaderContender contender = null;
-    private boolean hasLeadership = false;
-    private CompletableFuture<LeaderConnectionInfo> confirmationFuture = null;
-    private CompletableFuture<Void> startFuture = new CompletableFuture<>();
-    private UUID issuedLeaderSessionId = null;
-
-    /**
-     * Gets a future that completes when leadership is confirmed.
-     *
-     * <p>Note: the future is created upon calling {@link #isLeader(UUID)}.
-     */
-    public synchronized CompletableFuture<LeaderConnectionInfo> getConfirmationFuture() {
-        return confirmationFuture;
-    }
+    private final TestingLeaderElection startedLeaderElection = new TestingLeaderElection();
 
     @Override
-    public synchronized void start(LeaderContender contender) {
-        Preconditions.checkState(!getStartFuture().isDone());
-
-        this.contender = contender;
-
-        if (hasLeadership) {
-            contender.grantLeadership(issuedLeaderSessionId);
-        }
-
-        startFuture.complete(null);
+    public synchronized LeaderElection createLeaderElection() {
+        return startedLeaderElection;
     }
 
     @Override
     public synchronized void stop() throws Exception {
-        if (hasLeadership && contender != null) {
-            contender.revokeLeadership();
-        }
-
-        contender = null;
-        hasLeadership = false;
-        issuedLeaderSessionId = null;
-        startFuture.cancel(false);
-        startFuture = new CompletableFuture<>();
+        startedLeaderElection.triggerContenderCleanup();
     }
 
-    @Override
-    public synchronized void confirmLeadership(UUID leaderSessionID, String leaderAddress) {
-        if (confirmationFuture != null) {
-            confirmationFuture.complete(new LeaderConnectionInfo(leaderSessionID, leaderAddress));
-        }
+    public synchronized CompletableFuture<LeaderInformation> isLeader(UUID leaderSessionID) {
+        return startedLeaderElection.isLeader(leaderSessionID);
     }
 
-    @Override
-    public synchronized boolean hasLeadership(@Nonnull UUID leaderSessionId) {
-        return hasLeadership && leaderSessionId.equals(issuedLeaderSessionId);
+    public synchronized LeaderInformation isConfirmedLeader(UUID leaderSessionID) {
```

Comment:
The intention is good but I was a tad confused why isLeader and isConfirmedLeader return different types.

I feel like calling `isLeader.join()` explicitly in the test, or renaming this method to `isLeaderAndWaitForConfirm` may alleviate this.

Reply:
Fair enough, I guess calling `.join()` on the `CompletableFuture` is good enough. I'm gonna add JavaDoc to the `isLeader(UUID)` method, instead.